prov/sockets/src/sock_ep_msg.c

```diff
@@ -1236,3 +1256,158 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
 	return ret;
 }
 
+static void sock_ep_cm_check_msg_list(struct sock_ep_cm_head *cm_head)
+{
+	struct dlist_entry *entry;
+	struct sock_conn_req_handle *hreq;
+
+	fastlock_acquire(&cm_head->list_lock);
+	while (!dlist_empty(&cm_head->close_list)) {
+		entry = cm_head->close_list.next;
+		dlist_remove(entry);
+		hreq = container_of(entry, struct sock_conn_req_handle, entry);
+
+		sock_ep_cm_unmonitor_handle(hreq);
+		ofi_close_socket(hreq->sock_fd);
+		hreq->sock_fd = -1;
+
+		pthread_mutex_lock(&hreq->mutex_cond.mutex);
+		hreq->is_finalized = 1;
+		pthread_cond_signal(&hreq->mutex_cond.cond);
+		pthread_mutex_unlock(&hreq->mutex_cond.mutex);
+	}
+	fastlock_release(&cm_head->list_lock);
+}
+
+static void *sock_ep_cm_thread(void *arg)
+{
+	int num_fds, i;
+	struct sock_ep_cm_head *cm_head = arg;
+	void *ep_contexts[SOCK_EPOLL_WAIT_EVENTS];
+	struct sock_conn_req_handle *handle;
+	struct sock_conn_hdr msg;
+
+	while (cm_head->do_listen) {
+		sock_ep_cm_check_msg_list(cm_head);
+
+		num_fds = fi_epoll_wait(cm_head->emap, ep_contexts,
+		                        SOCK_EPOLL_WAIT_EVENTS, -1);
+		if (num_fds < 0) {
+			SOCK_LOG_ERROR("poll failed : %s\n", strerror(errno));
+			continue;
+		}
+
+		for (i = 0; i < num_fds; i++) {
+			handle = ep_contexts[i];
+
+			if (handle == NULL) { /* Signal event */
+				fd_signal_reset(&cm_head->signal);
+				continue;
+			}
+
+			if(sock_cm_recv(handle->sock_fd, &msg, sizeof(msg), MSG_PEEK)) {
+				SOCK_LOG_ERROR("io failed for fd %d\n", handle->sock_fd);
+				if (handle->ep) {
+					if (!handle->pep) /* Active-side */
+						sock_ep_cm_report_connect_fail(handle->ep, NULL, 0);
+					sock_ep_cm_shutdown_report(handle->ep, 0);
+				}
+
+				sock_ep_cm_unmonitor_handle(handle);
+				ofi_close_socket(handle->sock_fd);
+				handle->sock_fd = -1;
```

Comment:
This duplicates code from sock_ep_cm_check_msg_list().  Use a common function or move the handle to the close_list and process there.

The pairing of sock_ep_cm_unmonitor_handle() and ofi_close_socket() occurs a few times, but not always.  It would be nice if this can be made consistent.

Reply:
Done. I've moved `ofi_close_socket()` into `sock_ep_cm_unmonitor_handle()`, which simplifies the code.